arraycontext/container/arithmetic.py

```diff
@@ -268,18 +301,66 @@ def wrap(cls: Any) -> Any:
                     "'_deserialize_init_arrays_code'. If this is a dataclass, "
                     "use the 'dataclass_array_container' decorator first.")
 
+        if cls_has_array_context_attr is _FailSafe:
+            def actx_getter_code(arg: str) -> str:
+                return f"_get_actx({arg})"
+        else:
+            def actx_getter_code(arg: str) -> str:
+                return f"{arg}.array_context"
+
         from pytools.codegen import CodeGenerator, Indentation
         gen = CodeGenerator()
         gen("""
             from numbers import Number
             import numpy as np
-            from arraycontext import ArrayContainer
+            from arraycontext import (
+                ArrayContainer, get_container_context_recursively)
+            from warnings import warn
 
             def _raise_if_actx_none(actx):
                 if actx is None:
                     raise ValueError("array containers with frozen arrays "
                         "cannot be operated upon")
                 return actx
+
+            def _get_actx(ary):
+                try:
+                    return ary.array_context
+                except Exception as e:
+                    warn(f"Accessing '{type(ary).__name__}.array_context' failed "
+                        f"({type(e)}: {e}). This should not happen and is "
+                        "deprecated. "
+                        "Please fix the implementation of "
+                        f"'{type(ary).__name__}.array_context' "
+                        "and then set _cls_has_array_context_attr=True when "
+                        "calling with_container_arithmetic to avoid the run time "
+                        "cost of the check that gave you this warning. "
+                        "Using expensive recovery for now.",
+                        DeprecationWarning, stacklevel=3)
+
+                return get_container_context_recursively(ary)
```

Comment:
Would it make sense to cache this in a `self.__actx` or something so it doesn't have to redo `get_container_context_recursively` on every math operation?

(same for `_get_actx_array_types_failsafe`)

Reply:
Although that's the very pessimistic case, so might not be worth adding some more complexity in here..

Reply:
IMO: No. The slower and noisier this is, the better, because that means we can get rid of all this nonsense sooner. There's an easy way out. Just set `_cls_has_array_context_attr=True` and none of this ever happens.